pylabrobot/resources/plate_adapter.py

```diff
@@ -0,0 +1,175 @@
+        dx (float): The x-coordinate offset for well positioning.
+        dy (float): The y-coordinate offset for well positioning.
+        dz (float): The z-coordinate offset for well positioning, i.e. the outside-bottom
+          of a well.
+        adapter_item_dx (Literal[9.0, 4.5, 2.25], optional): The x-dimension spacing of
+          wells. Defaults to 9.0.
+        adapter_item_dy (Literal[9.0, 4.5, 2.25], optional): The y-dimension spacing of 
+        wells. Defaults to 9.0.
+        site_pedestal_z (Optional[float], optional): The z-coordinate of the site pedestal.
+          Defaults to None.
+        category (Optional[str], optional): The category of the PlateAdapter.
+          Defaults to "plate_adapter".
+        model (Optional[str], optional): The model of the PlateAdapter. Defaults to None.
+   
+  Examples:
+    1. Using a "magnetic rack" as a PlateAdapter:
+
+      Define a `MFXCarrier`, define a `MFXModule` for deep well plates,
+      assign the DWP module to the MFXCarrier:
+      >>> mfx_carrier_1 = MFX_CAR_L5_base(name='mfx_carrier_1')
+      >>> mfx_carrier_1[1] = mfx_dwp_module_1 = MFX_DWP_module(name="mfx_dwp_module_1")
+
+      Define Alpaqua magnet as a PlateAdapter & assign to mfx_dwp_module_1:
+      >>> Alpaqua_magnum_flx_1 = Alpaqua_magnum_flx(name='Alpaqua_magnum_flx_1')
+      >>> mfx_dwp_module_1.assign_child_resource(Alpaqua_magnum_flx_1)
+
+      Define Cos_96_Rd plate & assign to PlateAdapter:
+      >>> Cos96_plate_1 = Cos_96_Rd(name='Cos96_plate_1')
+      >>> Alpaqua_magnum_flx_1.assign_child_resource(Cos96_plate_1)
+  """
+
+  def __init__(
+    self,
+    name: str,
+    size_x: float, size_y: float, size_z: float,
+    dx: float, dy: float, dz: float,
+    adapter_hole_size_x: float,
+    adapter_hole_size_y: float,
+    site_pedestal_z: float,
+    adapter_hole_dx: float = 9.0,
+    adapter_hole_dy: float = 9.0,
+    category: Optional[str] = None,
+    model: Optional[str] = None
+    ):
+    if adapter_hole_dx not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dx must be one of 9.0, 4.5, or 2.25")
+    if adapter_hole_dy not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dy must be one of 9.0, 4.5, or 2.25")
+
+    super().__init__(name=name, size_x=size_x, size_y=size_y, size_z=size_z,
+      category=category or "plate_adapter", model=model)
+
+    self._child_resource_location = None
+    self._child_resource: Optional[Resource] = None
+    self.dx = dx
+    self.dy = dy
+    self.dz = dz
+    self.adapter_hole_size_x = adapter_hole_size_x
+    self.adapter_hole_size_y = adapter_hole_size_y
+    self.adapter_hole_size_z = self.get_size_z() - self.dz
+    self.adapter_hole_dx = adapter_hole_dx
+    self.adapter_hole_dy = adapter_hole_dy
+    self.site_pedestal_z = site_pedestal_z
+
+  @property
+  def child_resource_location(self) -> Optional[Coordinate]:
+    return self._child_resource_location
+
+  def assign_child_resource(
+    self,
+    resource: Plate,
```

Comment:
This leads to the type error. According to the "Liskov substitution principle", APIs cannot be more specialized/constrained for subclasses:

`PlateAdapter` is a `Resource` and so the following code should be valid:

```python
my_resource: Resource = PlateAdapter(...)
```

but then

```python
my_resource.assign_child_resource(non_plate)
```

would fail, even though `Resource.assign_child_resource` allows it.

In https://github.com/PyLabRobot/pylabrobot/pull/143/commits/d076b471a0df9c784c4cb10cf8323c215266b7e1#diff-c76c0cd2964c9c9ee4af709b9dd64d713a464a50ccede9f2c0a798c0f994bf70R200 I solved this with a dynamic type check, which isn't the cleanest, but works for now I guess 🤷‍♂️